### ISSUE TITLE
Support force destruction of DataStream private connections 

### ISSUE THREAD

**Issue by laksh-bhat** [hashicorp/terraform-provider-google#13054]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version
1.3.4

### Affected Resource(s)

* google_datastream_private_connection

### Terraform Configuration Files

```tf
resource "google_datastream_private_connection" "datastream_conn" {
  # This resource must have an IP range with a minimum CIDR block of /29
  # https://cloud.google.com/datastream/docs/create-a-private-connectivity-configuration
  display_name          = "datastream private connection"
  project                     = var.project
  location                   = var.region

  private_connection_id = "datastream-conn"

  vpc_peering_config {
    vpc    = var.private_network_id
    subnet = var.datastream_subnet_cidr
  }
}
```

### Debug Output

```
---[ REQUEST ]---------------------------------------
DELETE /v1/projects/relyance-internal/locations/us-central1/privateConnections/datastream-conn?alt=json HTTP/1.1
Host: datastream.googleapis.com
User-Agent: Terraform/1.3.4 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google/dev
Content-Type: application/json
Accept-Encoding: gzip

2022-11-16T11:05:26.577-0800 [INFO]  provider.terraform-provider-google_v4.43.0_x5: 2022/11/16 11:05:26 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 400 Bad Request
Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Wed, 16 Nov 2022 19:05:26 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "error": {
    "code": 400,
    "message": "Resource '\"projects/<internal>/locations/us-central1/privateConnections/datastream-conn\"' has nested resources. If the API supports cascading delete, set 'force' to true to delete it and its nested resources.",
    "status": "FAILED_PRECONDITION"
  }
}

```

### Expected Behavior

`terraform apply -var-file="$TFVARS_FILE" -destroy` is successful

### Actual Behavior

```
│ Error: Error when reading or editing PrivateConnection: googleapi: Error 400: Resource '"projects/<internal>/locations/us-central1/privateConnections/datastream-conn"' has nested resources. If the API supports cascading delete, set 'force' to true to delete it and its nested resources.

```

### Steps to Reproduce

1. `terraform apply -var-file="$TFVARS_FILE" -destroy`

### References

* #10810

**Comment by edwardmedia** [comment]
@laksh-bhat are you able to see if other tools (like gcloud) work as you expected?

**Comment by laksh-bhat** [comment]
Yes, `gcloud` works as expected.
```
gcloud datastream private-connections delete 'datastream-conn' --location=us-central1 --force --quiet

Deleted private_connection [datastream-conn].
```

**Comment by edwardmedia** [comment]
@laksh-bhat I do see `force` is available in below api but I am not sure if that should be implemented in the Terraform. 1) For use case, how do you manage those child routes? If they are created via Terraform, you better continue let Terraform to manage deletion. 2) Implementing `force` seems against the pattern, as Terraform is more declarative. 

https://cloud.google.com/datastream/docs/reference/rest/v1/projects.locations.privateConnections/delete

If you want this feature, you may file an [enhancement](https://github.com/hashicorp/terraform-provider-google/issues/new?assignees=&labels=enhancement&template=01_enhancement.md) and then it will be triaged. This is not a bug and I am closing it accordingly.

**Comment by laksh-bhat** [comment]
I understand this may seem like an anti-pattern for terraform but the `private connections` API doesn't allow us to manage the routes explicitly (you set a subnet and private VPC through a peering config).  
```tf
resource "google_datastream_private_connection" "private_vpc_datastream_conn" {
  # This resource must have an IP range with a minimum CIDR block of /29
  # https://cloud.google.com/datastream/docs/create-a-private-connectivity-configuration
  display_name          = "datastream private connection"
  project               = var.project
  location              = var.region
  private_connection_id = "${var.region}-${var.cell}-${var.env}-datastream-conn"

  vpc_peering_config {
    vpc    = var.private_network_id
    subnet = var.datastream_subnet_cidr
  }
}
```
The routes created by this, hence, aren't in terraform state. I strongly urge you to reconsider this decision to not implement the `force` flag. Without this flag, the resource is pretty much useless.

**Comment by melinath** [comment]
Note from triage: Reopening to track the request to add a field for configuring `force` to send to the API at deletion time. This should use a `deletion_policy` field - see https://googlecloudplatform.github.io/magic-modules/best-practices/deletion-behaviors/

**Comment by manitgupta** [comment]
I ran into this too. Is there an update on how to use terraform to delete `google_datastream_private_connection` resource? Without the ability to delete, the resource is pretty much useless.

**Comment by ota2000** [comment]
I'd like to work on this by adding a `deletion_policy` field as suggested by @melinath.

Before I start, I'd like to confirm the expected behavior for the default value.

Currently, PR #11394 hardcodes `force=true` on all deletions. Adding `deletion_policy` would replace this with user-controlled behavior, but the default value affects backward compatibility:

| Default | Behavior | Trade-off |
|---|---|---|
| `"FORCE"` | force=true (same as current) | Backward compatible, but implicit cascade delete |
| `""` (empty) | force=false | Safer, but breaks existing users who rely on current force=true behavior |

My suggestion is to default to `"FORCE"` to maintain backward compatibility with the current behavior introduced in #11394, and let users explicitly set `deletion_policy = ""` or remove it to opt out of force deletion.

Does this approach sound reasonable?

**Comment by melinath** [comment]
Correct, the deletion_policy field would need to retain backwards-compatibility with the current behavior; the default could be changed in a future major release. @NickElliot is there anything else to keep in mind?

**Comment by NickElliot** [comment]
as long as the field is a string and named "deletion_policy", it would be compatible!

**Comment by ota2000** [comment]
I've opened a PR to add the `deletion_policy` field: https://github.com/GoogleCloudPlatform/magic-modules/pull/17120

Thanks @melinath and @NickElliot for the guidance.
